lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php

```diff
@@ -263,41 +287,76 @@ private function getEntity(array $data, $dqlAlias)
         }
 
         if (isset($this->_hints[Query::HINT_REFRESH_ENTITY]) && isset($this->rootAliases[$dqlAlias])) {
-            $this->registerManaged($this->_metadataCache[$className], $this->_hints[Query::HINT_REFRESH_ENTITY], $data);
+            $this->registerManaged($this->_metadataCache->{$className}, $this->_hints[Query::HINT_REFRESH_ENTITY], $data);
         }
 
         $this->_hints['fetchAlias'] = $dqlAlias;
 
-        return $this->_uow->createEntity($className, $data, $this->_hints);
+        $managedEntity = ($this->entityTryGetCallbacksByEntityName->{$className})($data);
```

Comment:
Can you extract this part (till the end of the method) to a private method... it's quite confusing.

Reply:
No, I can't due to performance implications. A single method call here would already be extremely heavy.

Reply:
This additional call is already the reason why the performance degraded - I need to see if I can find a better way to get rid of it, or maybe I'll try optimising `buildEntityTryGet` or `UnitOfWork#createEntity()` (which is already optimised and horrible to the extreme)